pkg/models/registries/registry_client.go

```diff
@@ -0,0 +1,174 @@
+package registries
+
+import (
+	"compress/gzip"
+	"errors"
+	"fmt"
+	"github.com/docker/docker/api/types"
+	log "github.com/golang/glog"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+	"time"
+)
+
+const (
+	// DefaultDockerRegistry is the default docker registry address.
+	DefaultDockerRegistry = "https://registry-1.docker.io"
+
+	DefaultTimeout = 30 * time.Second
+)
+
+var (
+	bearerRegex = regexp.MustCompile(
+		`^\s*Bearer\s+(.*)$`)
+	basicRegex = regexp.MustCompile(`^\s*Basic\s+.*$`)
+
+	// ErrBasicAuth indicates that the repository requires basic rather than token authentication.
+	ErrBasicAuth = errors.New("basic auth required")
+
+	gcrMatcher = regexp.MustCompile(`https://([a-z]+\.|)gcr\.io/`)
+
+	reProtocol = regexp.MustCompile("^https?://")
+)
+
+// Registry defines the client for retrieving information from the registry API.
+type Registry struct {
+	URL      string
+	Domain   string
+	Username string
+	Password string
+	Client   *http.Client
+	Opt      Opt
+}
+
+// Opt holds the options for a new registry.
+type Opt struct {
+	Domain  string
+	Timeout time.Duration
+	Headers map[string]string
+	UseSSL  bool
+}
+
+type authToken struct {
+	Token       string `json:"token"`
+	AccessToken string `json:"access_token"`
+}
+
+type authService struct {
+	Realm   *url.URL
+	Service string
+	Scope   []string
+}
+
+func CreateRegistryClient(username, password, domain string) (*Registry, error) {
+	authDomain := domain
+	auth, err := GetAuthConfig(username, password, authDomain)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create the registry client.
+	log.Infof("domain: %s", domain)
+	log.Infof("server address: %s", auth.ServerAddress)
+
+	return New(auth, Opt{
+		Domain: domain,
+	})
+}
+
+// GetAuthConfig returns the docker registry AuthConfig.
+func GetAuthConfig(username, password, registry string) (types.AuthConfig, error) {
+	registry = setDefaultRegistry(registry)
+	if username != "" && password != "" && registry != "" {
+		return types.AuthConfig{
+			Username:      username,
+			Password:      password,
+			ServerAddress: registry,
+		}, nil
+	}
+
+	log.Info("Using registry ", registry, " with no authentication")
+
+	return types.AuthConfig{
+		ServerAddress: registry,
+	}, nil
+
+}
+
+func setDefaultRegistry(serverAddress string) string {
+	if serverAddress == "docker.io" || serverAddress == "" {
+		serverAddress = DefaultDockerRegistry
+	}
+
+	return serverAddress
+}
+
+func newFromTransport(auth types.AuthConfig, opt Opt) (*Registry, error) {
+	if len(opt.Domain) < 1 || opt.Domain == "docker.io" {
+		opt.Domain = auth.ServerAddress
+	}
+	url := strings.TrimSuffix(opt.Domain, "/")
+	authURL := strings.TrimSuffix(auth.ServerAddress, "/")
+
+	if !reProtocol.MatchString(url) {
```

Comment:
it's a little bit confused here to use regexp, you can just go with
```
if !strings.HasPrefix(url, "http://") && !strings.HasPrefix(url, "https://") 
```
even, you make this part as function,

```
completeURL = func (url string) string {
     if !strings.HasPrefix(url, "http://") && !strings.HasPrefix(url, "https://") {
           if opt.UseURL {
                 url = "http://" + url
           } else {
                  url = "https://" + url
           }
     }
    return url
}
url = completeURL(url)
authURL = completeURL(authURL)
```